### ISSUE TITLE
Build stand-alone Citadel

### ISSUE THREAD

**Issue by JimmyCYJ** [istio/istio#9177]
**Describe the feature request**
Remove the dependency on Kubernetes API server from Citadel.
Such stand-alone Citadel is needed by our new test framework, which is able to run every component in local environment as a process.

https://github.com/istio/istio/issues/9177

**Comment by JimmyCYJ** [comment]
cc @myidpt @hklai

**Comment by myidpt** [comment]
Thanks for creating this.

**Comment by myidpt** [comment]
Bump this to P0 in Q3.

**Comment by JimmyCYJ** [comment]
Discussed with @ozevren about Galley running in a native environment. Galley watches file system https://github.com/istio/istio/tree/master/galley/pkg/source for input config and output. 

Galley watches a folder and loads K8s resources when it receives a SIGHUP. Galley also has code to automatically watch folder for changes, but that caused issue with INODE exhaustion in the cluster. That is why SIGHUP mode is in use. 

Galley's implementation is not just for test, but also designed for real use cases such as VM deployment. 

For stand-alone Citadel, we need to design similar approach to load service accounts from folders on SIGHUP, and perhaps write secrets into folder for Envoy sidecars to pick up.

**Comment by myidpt** [comment]
Thanks Jimmy. I think we can potentially reuse https://github.com/istio/istio/tree/master/galley/pkg/source by moving that to istio/pkg. We can save a decent amount of work if Galley has already done similar things.

**Comment by ozevren** [comment]
Let's be careful here.
First, I am close to landing a refactoring of Galley in main (currently it is in collab-galley). Once it lands, I'll work to make it the default set of libaries Galley uses. it changes pkg/source. Any reorganization we need should take this into account.
https://github.com/istio/istio/tree/collab-galley/galley/pkg/config/source/kube

Second, istio/pkg is meant for top-level/general utilities, and not for our heavy machinery code. Besides, there is a desire to not have K8s dependencies from there. It is probably better to start another repo istio/config and move some of the stuff there.
@geeknoid

**Comment by myidpt** [comment]
Thanks for letting us know, Oz. I can understand the decision on istio/pkg usage, but I feel it's hard to sync the two repos if we put the common utils into a separate repo. Or maybe it's not as hard as I thought? :)

There was also an intent to put the `listwatch` (https://github.com/istio/istio/tree/master/security/pkg/listwatch) functionality to a common place. And I do see we already have https://github.com/istio/istio/tree/master/pkg/kube directory. I think it would be great if we can make the decision on where to host this code, as a later decision may cause more code refactoring.

**Comment by ozevren** [comment]
We *are* moving to a multirepo world. There is already a pkg repo where quite a few things have already moved to (not istio/pkg). Anything we put in istio/pkg will also need to move somewhere. The role of the pkg repo is to host core utility libraries with minimal dependencies. K8s dependencies are not desired in that repo. So, I think we need a separate repo for those.

I am landing the new libraries in master now:  https://github.com/istio/istio/pull/15891. Once this lands, you can take a dependency to this directly. As long as your dependencies are confined in pkg/config/... we can move the code appropriately.
